yare-model-validators/src/main/java/com/sabre/oss/yare/model/validator/ReferenceValidator.java

```diff
@@ -150,6 +159,33 @@ private void checkCollectionOperator(Type referenceType, String reference, Strin
         }
     }
 
+    private List<String> resolveNames(Rule rule) {
+        return Streams.concat(
+                rule.getAttributes().stream()
+                        .map(Attribute::getName),
+                rule.getFacts().stream()
+                        .map(Fact::getIdentifier)
+        ).collect(Collectors.toList());
+    }
+
+    private void checkNamesAreNotDuplicated(List<String> names, ValidationResults results) {
+        Set<String> duplicatedNames = names.stream()
+                .filter(i -> Collections.frequency(names, i) > 1)
+                .collect(Collectors.toSet());
+        if (!duplicatedNames.isEmpty()) {
+            append(results, ValidationResult.error("rule.ref.duplicated-names",
+                    "Naming Error: There are duplicated names -> " + duplicatedNames));
+
+        }
+    }
+
+    private void checkReservedNamesAreNotUsed(List<String> names, ValidationResults results) {
+        if (names.contains(CONTEXT)) {
+            append(results, ValidationResult.error("rule.ref.reserved-names",
+                    "Naming Error: Reserved names are used -> [ctx]"));
```

Comment:
Please use CONTEXT variable to describe error root cause. Just in case someone changes that value in the future.